### PR TITLE
Make functions without a return type explicit in the ast

### LINKED ISSUE #851: Make functions without a return type explicit in the ast
Currently we treat both function that has a return type `nil` (`function f() returns nil { `) and functions without a return type (`function f() {`) to be same in the ast. We need to make this difference explicit in the ast for #816

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Make functions without a return type explicit in the ast
Resolves #851

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front.syntax/tests/testPos.bal [hunk 1]
```diff
@@ -592,7 +591,11 @@ function testValidTypeDescEnd(SourceFile file, Position endPos, TypeDesc td) ret
 
 
 function typeDescContainRightSqureBracket(TypeDesc td) returns boolean {
```
Comment (jclark) [inline compiler/modules/front.syntax/tests/testPos.bal]: Missing an `s` and an `a`.

File: compiler/modules/front/resolveTypes.bal [hunk 2]
```diff
@@ -267,7 +268,8 @@ function resolveTypeDesc(ModuleSymbols mod, s:ModuleLevelDefn modDefn, int depth
                 a.push(arg.td);
             }
             t:SemType[] args = from var x in a select check resolveTypeDesc(mod, modDefn, depth + 1, x);
-            t:SemType ret = check resolveTypeDesc(mod, modDefn, depth + 1, td.ret);
+            s:TypeDesc? retTy = td.ret;
+            t:SemType ret = check resolveTypeDesc(mod, modDefn, depth + 1, retTy?: { startPos: td.endPos, endPos: td.endPos, builtinTypeName: "null" });
```
Comment (jclark) [inline compiler/modules/front/resolveTypes.bal]: If td.ret is (), then the SemType is t:NIL.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/tests/testPos.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/tests/testPos.bal]: fixed

File: compiler/modules/front/resolveTypes.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/resolveTypes.bal]: fixed

**Pull request merged**